### ISSUE TITLE
CPU performance on ARM processors

### ISSUE THREAD

**Issue by bigfatbrowncat** [Ryujinx/Ryujinx#1884]
## Feature Request

_(I have no idea how the code translation in Ryujinx works right now, please let me know if this already works)_

Obviously, the weakest point of the emulator itself is the necessity of the dynamic code translation.

Meanwhile there are many ARM SoC-based computers out there right now. And some of them are pretty powerful. I'll list two of them:
1. Nvidia Jetson family - the hardware of this machine is the next generation of the chip that is used in the Switch itself, I presume.
2. The new Apple Mac family, ARM-based. They have custom chips (not the Nvidia ones), but they will still be ARM and have high potential...

#### Overview:
- The basic idea is to support "direct" execution of the simple ARM instructions to make the code work faster on the "native" platform. That does not include any Switch-special devices as they will still have to be simulated.

#### Smaller Details: 
- Unfortunately I have no idea about the emulator architecture, so I can't be more specific. My suggestion is to support ARM on ARM as directly as possible.

#### Nature of Request:  
- Addition
  - Another dynamic instruction translation module that works like VMWare or VirtualBox without VT-d - just sandboxing the code instead of translating it.

### Why would this feature be useful?
Happy Nvidia Jetson buyers, who will play Nintendo Switch games on nearly-native performance forever (even after the Nin* company will drop the support).
Happy Apple computer (MacBook Pro or Mac Mini) users, who will _\<the same\>_.

**Comment by jduncanator** [comment]
Thanks for your request!

One of the big limiting factors of running Ryujinx on ARM-based SBC's and devices is the RAM requirements. Ryujinx really needs, at a minimum, 8GB of RAM, where most ARM-based devices on the market ship with a maximum of 4GB. We've only recently seen the addition of 8GB SKUs coming to market in the forms of things like the Raspberry Pi 4.

Unfortunately, even with direct ARM execution, the NVIDIA Jetson, as well as the NVIDIA SHIELD and other "embedded" devices lack enough memory to properly emulate the Switch (at least at the current time). The most obvious target is, as you point out, the new M1 based Apple devices.

Rest assured, this is in our long term plan. Initially, we'll be bringing up ARM support as a dynamic code translation target architecture in the short term, as this is a relatively straightforward thing to add to our existing JIT. This work is already underway. Once we have this working, I imagine a natural extension to this will be running the code in a sandbox.

Keep an eye out on Patreon posts, and our change log, for more information on our progress on this!

**Comment by bigfatbrowncat** [comment]
@jduncanator, I understand that it has to have enough memory. The new NVidia board, [Jetson Xavier NX](https://developer.nvidia.com/embedded/jetson-xavier-nx) has exactly 8GB for the price of $400

**Comment by genos1213** [comment]
Aren't the Nvidia Jetsons pretty low spec? I don't think any existing one would ever be able to emulate the Switch well. Meanwhile flagship Android smartphones tend to ship with 8GB RAM with the latest ARM consumer CPUs.

The Jetson Xavier NX has a CPU similar in performance to a Snapdragon 845 from 2018. In contrast the M1 probably has the CPU performance of a 2023 Snapdragon flagship phone.

**Comment by bigfatbrowncat** [comment]
@genos1213, look, If I understand correctly, NVidia Jetson TX1 is the Tegra X1 board. Look: https://www.anandtech.com/show/9779/nvidia-announces-jetson-tx1-tegra-x1-module-development-kit

X1 is the hardware that the original Switch is based on, right?

And TX2 is the advanced version of TX1 and NX is the new model of this arch.

Please, correct me if I'm wrong. As far as I know, Xavier is the most powerfull ARM-based gaming machine so far.

This is how it runs Dolphin, for instance:
https://www.youtube.com/watch?v=nijz19dhZgg

**Comment by bylaws** [comment]
Yeah the NX is for sure powerful enough, skyline is like 2x faster than a switch with NCE on an 845, so even 50% native perf should be okay. Its gpu is also a beast, it wouldn't need vulkan as it has desktop OGL.

**Comment by ExE-Boss** [comment]
There are also ARM‑based **Windows 10** devices.

**Comment by CeruleanSky** [comment]
Dolphin has achieved native M1 Support "[Temptation of the Apple: Dolphin on macOS M1](https://dolphin-emu.org/blog/2021/05/24/temptation-of-the-apple-dolphin-on-macos-m1/)" and perhaps some of the credited coders in the announcement can be contacted for advice if needed.

They pointed to this article on the Apple Developer website for tips ["Porting Just-In-Time Compilers to Apple Silicon"](https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon).

**Comment by jduncanator** [comment]
Dolphin already had a relatively feature complete ARM backend for their JIT used on a number of other platforms, so adding M1 as a target is a relatively straightforward procedure. Our JIT currently only targets x86, and while it was built to allow flexibility in extending it to additional backend architectures, there is still much work to be done to bring up this functionality.

This is still a long term goal, however is currently on the backburner in lieu of more fundamental features.

**Comment by edisionnano** [comment]
Dolphin also emulates a PowerPC CPU wile the Switch has an ARM64 one which would be more efficient to not emulate it on an ARM64 host (at least not fully). So I don't believe this is helpful.

**Comment by jduncanator** [comment]
Yea, the ultimate endgame would be to just run the original code directly on the target architecture, but this in and of itself is a massive change to the architecture of Ryujinx. The initial implementation will be ARMv8-on-ARMv8, as this is the path to least resistance.

**Comment by WalidCoding** [comment]
> 
> Rest assured, this is in our long term plan. Initially, we'll be bringing up ARM support as a dynamic code translation target architecture in the short term, as this is a relatively straightforward thing to add to our existing JIT. This work is already underway. Once we have this working, I imagine a natural extension to this will be running the code in a sandbox.
> 
> Keep an eye out on Patreon posts, and our change log, for more information on our progress on this!

@jduncanator its already 1 year later. Is this short-term goal (armv8-on-armv8) in process? I couldnt find any info on this.

**Comment by holaimscott** [comment]
@jduncanator 

Yeah, I wanted to ask about ARM support as well, specially with the increase of ARM devices and [Windows ARM support](https://www.youtube.com/watch?v=6NADjIc9uNk) [coming up](https://www.windowscentral.com/snapdragon-8cx-gen-3-leaked-benchmarks-look-good-news-windows-arm-pcs), it would be so cool to see your team consider making a native ARM port as well.

Mainly because this could help set the foundation for future ARM devices that might be more powerful than our current ones. Other emulators have done this and has helped them be future-proof/ accessible to different platforms within windows, [like Duckstation, PPSSPP, and Dolphin](https://www.youtube.com/watch?v=93grsZl3lHo)

Of course, it's easier said than done and I recognize a lot of hard work comes into place.

**Comment by drunlade** [comment]
Having picked up a shiny new RK3588S (the 8GB RAM version, though notably it goes up to 32GB) dev board I find my self interested in the Arm variants of various emulators, and quite specifically Ryujinx.

**Comment by gdkchan** [comment]
Switch emulation also requires a good GPU, it's not just about the CPU and memory.

**Comment by drunlade** [comment]
I understand, and the RK3588S has a pretty decent GPU, it's a massive step up over previous SoC that I've used or seen. How much GPU overhead vs real switch hardware is needed?

`Mali-G610 MP4 “Odin” GPU with support for OpenGLES 1.1, 2.0, and 3.2, OpenCL up to 2.2 and Vulkan1.2 - Performance FP32 (float) -> 610.6 GFLOPS`

**Comment by jduncanator** [comment]
I guess what gdk should have said was a good GPU _driver_, a lot of these devices have theoretically decent performance, but are let down by hugely inadequate drivers filled with bugs and lacking in performance. I don't see why the RK3588 would change what's been the (unfortunate) status-quo in the ARM SBC sector for the last decade.

Until ARM decide to release generic drivers, that work across all incarnations of their Mali GPUs, and we no longer need to rely on the SoC or SBC vendor to provide driver updates, it's unlikely emulating the Switch on an SBC will be feasible anytime soon.

**Comment by drunlade** [comment]
That's a fair point on the drivers, I didn't really consider that side. I guess I'll see what I get out of it. Early gaming benchmarks show it working pretty well, but that's subjective (and also running Android).

Regarding GPU overhead, I guessed that'd be the sticking point. I was under the impression switch topped out at around 500 while docked. Has that changed? Even if it hasn't it's still not hitting the 2x. I guess I need to set my expectations/hopes lower!

**Comment by bylaws** [comment]
Mali drivers are super terrible in general, if you had an adreno board you'd probably have a chance at running most things given you did some GPU optimisations to avoid inline updates

**Comment by drunlade** [comment]
However there are open source drivers now that "allegedly" are good / much better, possible the old "Mali drivers suck" issue might be less so.

**Comment by drunlade** [comment]
Looks like Skyline is already running pretty well on RK3588S based boards, that's the Android based switch emulator, so perhaps not so bad after all.

**Comment by notpushkin** [comment]
Congrats on the launch! 👏  https://blog.ryujinx.org/the-impossible-port-macos/

Sorry for (maybe) hijacking this issue, quick question: how hard would it be to get ARM-optimized builds on [Asahi Linux](https://asahilinux.org/) (and other ARM varieties running Linux, too)?

**Comment by edisionnano** [comment]
I second this, windows and linux arm builds would be great. I wanna see Ryujinx running on a nintendo switch with Linux 4 Tegra, arm SBCs and even android phones

**Comment by gdkchan** [comment]
It should work on Arm Linux once the Arm64 JIT is merged. I'm not sure if we will be able to run the Arm code directly though. Maybe using KVM it's possible, but I haven't tried it yet. I already got games running on the Raspberry Pi 4 with JIT and "NCE", but performance was extremely poor due to the very limited GPU that the Pi has.

I don't currently own a Windows Arm device so support for that one will likely depend on someone else.

**Comment by driver1998** [comment]
I would like to tackle the Windows ARM port but currently both Qualcomm and Parallels on M1 does not support Vulkan, and only do OpenGL 3.3. (Qualcomm: D3D12 Native, OpenGL 3.3 over GLonD3D12, Parallels: D3D11 and OpenGL 3.3)

Looks like Ryujinx needs at least OpenGL 4.3, I think we are out of luck.

**Comment by Quenii** [comment]
> Aren't the Nvidia Jetsons pretty low spec? I don't think any existing one would ever be able to emulate the Switch well. Meanwhile flagship Android smartphones tend to ship with 8GB RAM with the latest ARM consumer CPUs.
> 
> The Jetson Xavier NX has a CPU similar in performance to a Snapdragon 845 from 2018. In contrast the M1 probably has the CPU performance of a 2023 Snapdragon flagship phone.

https://wiki.radxa.com/Rock5/5b
this SBC got 4 core ARM64  running at 2.4GHz, and 8GB RAM

so do you have plan to run natively on such platform?

**Comment by marysaka** [comment]
> > Aren't the Nvidia Jetsons pretty low spec? I don't think any existing one would ever be able to emulate the Switch well. Meanwhile flagship Android smartphones tend to ship with 8GB RAM with the latest ARM consumer CPUs.
> > The Jetson Xavier NX has a CPU similar in performance to a Snapdragon 845 from 2018. In contrast the M1 probably has the CPU performance of a 2023 Snapdragon flagship phone.
> 
> https://wiki.radxa.com/Rock5/5b this SBC got 4 core ARM64 running at 2.4GHz, and 8GB RAM
> 
> so do you have plan to run natively on such platform?

Don't get your hopes up considering the GPU is ARM Mali based.

**Comment by Quenii** [comment]
> > > Aren't the Nvidia Jetsons pretty low spec? I don't think any existing one would ever be able to emulate the Switch well. Meanwhile flagship Android smartphones tend to ship with 8GB RAM with the latest ARM consumer CPUs.
> > > The Jetson Xavier NX has a CPU similar in performance to a Snapdragon 845 from 2018. In contrast the M1 probably has the CPU performance of a 2023 Snapdragon flagship phone.
> > 
> > 
> > https://wiki.radxa.com/Rock5/5b this SBC got 4 core ARM64 running at 2.4GHz, and 8GB RAM
> > so do you have plan to run natively on such platform?
> 
> Don't get your hopes up considering the GPU is ARM Mali based.

How about run cpu code natively, while gpu thru emulation?

**Comment by drunlade** [comment]
The Mesa project is, apparently, adding support for the line of GPUs found in the RK3588 series of SoC (as found in the Rock5b and other recent boards), which hopefully includes Vulkan support, since the SoC is literally advertised as supporting it. Also of note is the fact that Skyline (Android based Switch emulator) runs "ok" on this SoC using the stock Android image provided for the board. However, it's safe to say that supporting this board is likely of low priority, as cool as it would be to see happen.

**Comment by LAP87** [comment]
I just got my Orange Pi 5 delivered, 8gb ram variant.
Sitting here reading up on what is and isn't supported regarding RK3588S, i hope to see more stuff being ported for arm, because this is one performing little package, 8k output hdmi, several hundred fps measured in 3D benchmarks

**Comment by jduncanator** [comment]
> I just got my Orange Pi 5 delivered, 8gb ram variant.
> 
> Sitting here reading up on what is and isn't supported regarding RK3588S, i hope to see more stuff being ported for arm, because this is one performing little package, 8k output hdmi, several hundred fps measured in 3D benchmarks

Nothings changed from the previous statements regarding the RK3588. The GPU is Mali based, and the drivers shipped with the OEM's proprietary, blob ridden BSP is over 2 years old.

**Comment by gdkchan** [comment]
We are working on Arm support now, so there will probably be some developments on that front in the following weeks.

**Comment by LAP87** [comment]
> 

Lovely to hear that, i've been having fun using it as a temporary plex media server with Armbian as the OS, so far it doesn't use HW acceleration for HEVC/H.265, AV1 or VP9 (outside of chromium).
But i look forward to how this will develop over the year to come :)

**Comment by eslerm** [comment]
On Ubunut Asahi I installed [Microsoft's .NET SDKs for arm64](https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-6.0.413-linux-arm64-binaries) to compile Ryujinx.

I was not able to compile the current release or the commit which introduced dotnet 7. Builds fails due to an illegal instruction.

The commit prior to dotnet 7's introduction compiles with dotnet 6. The GUI loads, but firmware does not. Loading firmwares errors with "Invalid byte sequence in conversion input". That might have been fixed during a dotnet 7 patch (or I might have made a mistake).

**Comment by Mis012** [comment]
Possible users for KVM support (apart from Linux on aarch64 Macs) would be chromebooks with qcom SoCs. They run Linux in EL2 (as opposed to most devices with qcom SoCs), they use Mesa even on stock afaik, and google upstreams support for the boards to Linux mainline so one can have a completely sane environment.

The switch itself (with L4T) would absolutely be a great target, though sadly it only has 4GB of RAM and idk if the memory use would necessarily get reduced by half with JIT out of the way

**Comment by eglauko** [comment]
Ryujinx can run on Retroid Pocker 4 Pro?

**Comment by gdkchan** [comment]
Retroid Pocket 4 Pro runs Android, and the emulator does not support Android yet.

**Comment by matalama80td3l** [comment]
I've got a arm64 m-atx mother board, which has a 8-core 2.6ghz(performance between A75 and A76) SOC. It is SBSA architecture, support kvm and has standard pcie(plug any dgpu supported by linux mainline). I have tested a amd r7-340 on it.  

So will we have a arm64-linux release that use KVM over JIT? I can do some test work. I think it is really cool though I know it's quite niche😥

**Comment by MetrosexualGarbodor** [comment]
The original request was basically asking for Nvidia Jetson and ARM Mac compatibility. Only one of those has been achieved, so I don't think it should be closed.

**Comment by eslerm** [comment]
Could a new issue be created for Nvidia specifically?

I read the title and initial comment as ARM generally and don't think that this issue should remain open until every ARM chip is supported.

**Comment by gdkchan** [comment]
> The original request was basically asking for Nvidia Jetson

If it is running Linux, then it should work in theory. I don't think anyone tested it, though.
